web/html/src/components/tree.js

```diff
@@ -0,0 +1,163 @@
+// @flow
+
+const { hot } = require('react-hot-loader');
+const React = require('react');
+const { CustomDiv } = require('components/custom-objects');
+
+type TreeItem = {
+  id: string,
+  data?: Object,
+  children?: Array<string>,
+};
+
+type TreeData = {
+  rootId: string,
+  items: Array<TreeItem>,
+};
+
+type Props = {
+  data: TreeData,
+  renderItem: (item: TreeItem, renderNameColumn: Function) => void,
+  header?: React.Node,
+  initiallyExpanded?: Array<string>,
+  onItemSelectionChanged?: (item: TreeItem, checked: boolean) => void,
+  initiallySelected?: Array<string>,
+};
+
+type State = {
+  visibleSublists: Array<string>,
+  selected: Array<string>,
+};
+
+class Tree extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      visibleSublists: this.props.initiallyExpanded || [],
+      selected: this.props.initiallySelected || [],
+    };
+  }
+
+  isSublistVisible = (id: string): boolean => {
+    return this.state.visibleSublists.indexOf(id) !== -1;
+  }
+
+  isSelected = (id: string): boolean => {
+    return this.state.selected.indexOf(id) !== -1;
+  }
+
+  handleVisibleSublist = (id: string): void => {
+    this.setState(oldState => {
+      const { visibleSublists } = oldState;
+      return {
+        visibleSublists: visibleSublists.indexOf(id) !== -1
+          ? visibleSublists.filter(item => item !== id)
+          : visibleSublists.concat([id]),
+      };
+    });
+  }
+
+  handleSelectionChange = (changeEvent: Event): void => {
+    if (changeEvent.target instanceof HTMLInputElement) {
+      const { value: id, checked } = changeEvent.target;
+      this.setState(oldState => {
```

Comment:
as we talked we can simplify this a lot using React-Hooks and user-immer :+1:  I will handle it tomorrow